### ISSUE TITLE
New metrics: Mean Directional Accuracy & Logarithm of the Accuracy Ratio

### ISSUE THREAD

**Issue by Garve** [alan-turing-institute/sktime#718]
Hi!

How about some time series regression metrics which are not present in scikit-learn? So far, scikit-learn has the MAPE. This measure has some serious shortcomings, see [here](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error#Issues). If you are interested in this topic, the following could be good additions:

## Mean directional accuracy
I think this one is especially interesting when you only want to predict the direction of your time series correctly. Up or down nothing else. A classic problem to apply this is when dealing with the prediction of stock prices. I don't really care if my model can tell me if the stock price rises by 10% or 20%, it's important that it rises. And the model won't output accurate percentages anyway. It makes it potentially easier to design a well-performing model according to this metric.

## Logarithm of the accuracy ratio
[Here](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=2635088), the author explains why it is a good measure. What I like about this measure: Usually, the MAPE punishes over and underestimating differently. If the actual value A is 2 and the forecast F is 1, the APE here is 0.5. If A=2 and F=4, the APE is 1. Two different scores, although the prediction is half or double the actual value, and I would like these errors to be the same. The Logarithm of the accuracy ratio does exactly that (0.48 in both cases). (By the way, the Mean directional accuracy does this, too.)

Another nice property is that the best constant prediction for some label vector y is the geometric mean of y. This is analog to the mean, which is the best constant prediction for the MSE metric, or the median for the MAD metric.

[You can find my implementations here.](https://github.com/Garve/scikit-bonus/blob/master/skbonus/metrics/_regression.py)

If you like, I can do a pull request.

Best

Robert

**Comment by fkiraly** [comment]
Brilliant, that would be nice!

Before you open a PR, there is already one on metrics, #672, and an issue, #671. Your metrics are not included as far as I can tell, but they would be nice additions.

However, since there was a bit of design discussion on the metrics interface (related to metrics that need training data), I would suggest to align with @RNKuhns and the PR to ensure we end up with a single metrics interface.


Other notes:
* An interesting (other) discussion would also be whether we should make metrics composable, e.g., take any classification metric, and turn it into an up/down metric, or an up/down/mid metric as yours are. Where the user could select thresholds between up/down/mid.
* should your second metric be the mean/average of logarithmic percentage errors, or the logarithm of mean percentage errors? If it is the first (as it seems), that's the same as the mean absolute error of forecasting the logarithmic series (because a logarithm of a fraction is a difference of logarithms). In that case, would it not be better to first take the logarithms of your series and just use MAE?

**Comment by RNKuhns** [comment]
@Garve We have been discussing the interface for forecasting metrics in #672 and #712. Both of these could potentially be added since both just depend on the true and predicted data (along with optional weights and multioutput values like the propsed API in #672).

We are close to committing #672 and these could probably be added as a pull request on top of that once it is finished. 

One question I had about the log-quotient metric is the squaring of the quotient. I was scanning through the source paper and missed the squared term there. It seems like this would cause the metric to penalize errors when the prediction is close to zero (so the quotient is close to zero) more rapidly than when the the error is due to the y_pred being > y_true. 

![image](https://user-images.githubusercontent.com/26907244/110265303-43392e00-7f89-11eb-8d12-35baaf4b83c6.png)


In terms of code, at first glance it looks like it matches up pretty much exactly with what I've got in the other (soon to be) new metrics (this makes sense since we are both following the style of scikit-learn's regression metrics). So I think the log-quotient metric can probably be added with minor tweaks to handle cases where the true valeus are zero and also to the docstrings (including some additional information on use case the values the metric can take, etc. ) as long as @fkiraly and @mloning  think it makes sense. 

The other main thing I'd note on my end, would be in the docstring it would be helpful to have a description of when users would want to use this metrics since other metrics like MASE also make up for some issues with MAPE and SMAPE and also work well for intermittent demand (don't have divide by zero issue when y_true = 0). 

On the mean directional accuracy, and directional accuracy, @fkiraly makes a good point on potentially generalizing the directional accuracy metric. I think having a mean_directional_accuracy function is useful, but is also a specific case.

[Pesaran and Timmerman (2009)](http://www.econ.cam.ac.uk/people-files/emeritus/mhp1/pp09/PesaranTimmermannjasa.2009.pdf) discuss directional accuracy as a multicategory generalization of binary directional accuracy. 

The (Blaskowitz and Herwartz) referenced in wikipedia entry for mean directional accuracy is also a interesting discussion of accounting for economic cost of directional accuracy (not just 1 or 0). 

Maybe we could have a more general function that is used in the calculation of mean_directional_accuracy? That way users could choose different options and create their own scoring functions based on economic cost of incorrect directional accuracy. 

I can post a potential function definition tomorrow.

**Comment by Garve** [comment]
Hi again!

Thanks for the feedback :)

> Other notes:
> 
>     * An interesting (other) discussion would also be whether we should make metrics composable, e.g., take any classification metric, and turn it into an up/down metric, or an up/down/mid metric as yours are. Where the user could select thresholds between up/down/mid.

That's indeed interesting! Do you know a good use-case for a "mean directional f1 score" for example?

 
>     * should your second metric be the mean/average of logarithmic percentage errors, or the logarithm of mean percentage errors? If it is the first (as it seems), that's the same as the mean absolute error of forecasting the logarithmic series (because a logarithm of a fraction is a difference of logarithms). In that case, would it not be better to first take the logarithms of your series and just use MAE?

Yeah, one could do that. But one could also argue that the MAPE is just the MAE + samples weights np.abs(1/y_true). And still, it is offered as a separate metric. I understand your point, of course :) it's a design choice about how far to go to make something more accessible to people that don't know the mathematical connections. I'm fine with whatever you people say since I'm still very new here.

**Comment by fkiraly** [comment]
@Garve, @RNKuhns, your discussion to me also highlights one additional thing:

We (will soon) have a lot of metrics implemented, but it is unlikely entirely clear to the user what all of these are. E.g., do you take log or fraction before or after taking the average, which average or of what the average is meant (squared or unsquared), and so on.

I do think we need a piece of documentation, e.g., markdown with TeX, that for each metric name or acronym fully spells out the definition in formulae, say in a set-up where you have forecasts $\widehat{y}_1,\dots, \widehat{y}_N$ and observations $y_1,\dots, y_N$.

Optimally, this comes with:
* a statement what is the best uninformed predictor, i.e., constant prediction, in i.i.d. setting, and/or in the forecasting setting (if applicable, e.g., 1-step)
* a taxonomy of metrics, e.g., separating genuine forecasting metrics from up/down or directional metrics

What do you think?

**Comment by fkiraly** [comment]
>I do think we need a piece of documentation

I can also help with this, just let me know

**Comment by RNKuhns** [comment]
I agree on the documentation.  To me it seems like we could add it in the user guide: https://www.sktime.org/en/latest/user_guide/forecasting.html.

**Comment by RNKuhns** [comment]
@garve and @fkiraly how does something like this work for proceeding on this:

Assuming we wrap up pull request #672 and issue #712, adding the logarithm of accuracy ratio metric is straight-forward, so I'd lean to it potentially being its own pull request. 

Then break out the discussion of directional accuracy metrics and documentation into seperate issues/pull requests. What are your thoughts?

**Comment by RNKuhns** [comment]
On the directional accuracy part, I think that mean_directional_accuracy works as a metric (and is used in articles), but has its limitations just like the magnitude based metrics.

In our documentation we'd probably want to capture something like the following:
- Directional accuracy is important to consider because a forecast with a low (good) forecasting metric value may not predict directional changes and lead to bad economic decisions (buying or selling stock at in opportune time or making bad inventory decisions based on demand forecasts)
- Simple approach to directional accuracy is mean directional accuracy but this also has limitations. 
    - Getting direction correct but having forecasts that are far from the true values can also lead to bad economic decisions (e.g. picking the wrong stock to invest in when basing your decision on forecasts of many stock prices)

The (Blaskowitz and Herwartz) paper [https://edoc.hu-berlin.de/bitstream/handle/18452/4869/52.pdf?sequence=1&isAllowed=y] discusses economic value of directional forecasts. I think we could try to accomodate something like that. 

That paper differentiates between economic value from 4 potential directional accuracy outcomes (i.e.: h_uu where true value went up and prediction went up, h_dd where both went down, h_du where the true value went down and the prediction up, and h_ud where the true value went up and prediction went down).

Could we have a function that had keyword arguments for `h_uu`, `h_dd`, `h_ud`, `h_du` that accepted numbers or callables (callables would have to take only `y_true`, `y_pred` and `y_train` as arguments. 

Note that I'm passing `y_train` because the Blaskowitz and Herwartz paper shows economic value as function of Y_t (last value of training set) and Y_{t+h}, so we'd probably want the functions to accommodate that.

```python
def mean_directional_value(y_true, y_pred, y_train, h_uu = 1, h_dd = 1, h_du = 0, h_ud = 0):
    pass
```

This would let `mean_directional_accuracy` be defined as:

```python
def mean_directional_accuracy(y_true, y_pred, y_train, h_uu = 1, h_dd = 1, h_du = 0, h_ud = 0):
    return mean_directional_value(y_true, y_pred, y_train)
```

**Comment by fkiraly** [comment]
>Then break out the discussion of directional accuracy metrics and documentation into seperate issues/pull requests. What are your thoughts?

Agreed - I think there should be issues/PR:
* PR with "standard" forecasting metrics
* design for directional metrics - maybe a phone call to align (us three + @mloning ?)
* PR for directionam metrics, conditional on design done
* documentation for all metrics; recommend this to be its own page, but tbd

**Comment by fkiraly** [comment]
@RNKuhns, @Garve, would you like to ping me (plus each other) on the sktime slack for a quick align call for how to split the work and PR, plus possible design discussions? @mloning is on a short thesis writing sabattical but will be catching up.

**Comment by mloning** [comment]
Hi @Garve take a look at #745 and let us know if that would be useful for you!
